halo2-base/src/virtual_region/copy_constraints.rs

```diff
@@ -104,6 +104,22 @@ impl<F: Field + Ord> CopyConstraintManager<F> {
         self.external_cell_count = 0;
         self.assigned.take();
     }
+
+    /// Constrains that `virtual` is equal to `external`. The `virtual` cell must have
+    /// already been raw assigned. This should only be called when `witness_gen_only` is false, otherwise it will panic.
+    ///
+    /// ## Panics
+    /// If witness generation only mode is true.
+    pub fn constrain_virtual_equals_external(
```

Comment:
This doesn't change the state of `CopyConstraintManager` so it should not use `&mut self`. Also this causes some side effects in `region`. Maybe move this into a helper function instead of a member function?

Reply:
OK I was debating where to put it, your suggestion makes sense, I'll move it out as a helper function.